src/main/java/org/jboss/netty/handler/execution/FairOrderedMemoryAwareThreadPoolExecutor.java

```diff
@@ -0,0 +1,259 @@
+    }
+
+    /**
+     * Executes the specified task concurrently while maintaining the event order.
+     */
+    @Override
+    protected void doExecute(Runnable task) {
+        if (task instanceof ChannelEventRunnable) {
+            ChannelEventRunnable eventRunnable = (ChannelEventRunnable) task;
+            EventTask newEventTask = new EventTask(eventRunnable);
+
+      /*
+       * e.g. Three event "Channel A (Event A1)","Channel A (Event A2)","Channel A (Event A3)" are
+       * submitted in sequence, then key "Channel A" is refer to the value of "Event A3", and there
+       * is a linked list: "Event A3" -> "Event A2" -> "Event A1" ( linked by the field "next" in
+       * EventTask )
+       */
+
+            Object key = getKey(eventRunnable.getEvent());
+            EventTask previousEventTask = map.put(key, newEventTask);
+
+            // Remove the entry when the channel closes.
+            removeIfClosed(eventRunnable, key);
+
+            // try to setup "previousEventTask -> newEventTask"
+            // if success, then "newEventTask" will be invoke by "previousEventTask"
+            if (previousEventTask != null) {
+                if (compareAndSetNext(previousEventTask, null, newEventTask)) {
+                    return;
+                }
+            }
+
+            // Two situation here:
+            // 1. "newEventTask" is the header of linked list
+            // 2. the "previousEventTask.next" is already END
+            // At these two situations above, just execute "newEventTask" immediately
+            doUnorderedExecute(newEventTask);
+
+            // Remove the entry when the channel closes.
+            ChannelEvent event = eventRunnable.getEvent();
+            if (event instanceof ChannelStateEvent) {
+                Channel channel = event.getChannel();
+                ChannelStateEvent se = (ChannelStateEvent) event;
+                if (se.getState() == ChannelState.OPEN && !channel.isOpen()) {
+                    removeKey(key);
+                }
+            }
+        } else {
+            doUnorderedExecute(task);
+        }
+    }
+
+    private void removeIfClosed(ChannelEventRunnable eventRunnable, Object key) {
+        ChannelEvent event = eventRunnable.getEvent();
+        if (event instanceof ChannelStateEvent) {
+            Channel channel = event.getChannel();
+            ChannelStateEvent se = (ChannelStateEvent) event;
+            if (se.getState() == ChannelState.OPEN && !channel.isOpen()) {
+                removeKey(key);
+            }
+        }
+    }
+
+    /**
+     * call removeKey(Object key) when the life cycle of the key ends, such as when the channel is closed
+     */
+    protected boolean removeKey(Object key) {
+        return map.remove(key) != null;
+    }
+
+    protected Object getKey(ChannelEvent e) {
+        return e.getChannel();
+    }
+
+    @Override
+    protected boolean shouldCount(Runnable task) {
+        //noinspection SimplifiableIfStatement
```

Comment:
It is the same as class OrderedMemoryAwareThreadPoolExecutor, so i did not take notice of it, whatever, i will fix it  ;-)